playerctl/playerctl-cli.c

```diff
@@ -16,279 +16,359 @@
-
-    g_print("%s", value);
-
-    g_free(value);
-  } else if (g_strcmp0(command[0], "status") == 0) {
-    /* STATUS */
-    gchar *status = NULL;
-    g_object_get(player, "status", &status, NULL);
-
-    if (status) {
-      g_print("%s\n", status);
+  } else {
+    g_object_get(player, "position", &offset, NULL);
+    g_print("%g\n", (double)offset / 1000000);
+  }
+
+  return TRUE;
+}
+
+static gboolean set_or_get_volume (const gchar *value, PlayerctlPlayer *player, GError **error)
+{
+  const gchar *volume = value;
+  gdouble level;
+
+  if (volume) {
+    size_t last = strlen(volume) - 1;
+
+    if(volume[last] == '+' || volume[last] == '-') {
+      gdouble adjustment = g_ascii_strtod(volume, NULL);
+
+      if(volume[last] == '-') {
+          adjustment *= -1;
+      }
+
+      g_object_get(player, "volume", &level, NULL);
+      level += adjustment;
     } else {
-      g_print("Not available\n");
+      level = g_ascii_strtod(volume, NULL);
     }
-
-    g_free(status);
+    g_object_set(player, "volume", level, NULL);
   } else {
-    /* unrecognized command */
-    g_print("%s", g_option_context_get_help(context, TRUE, NULL));
+    g_object_get(player, "volume", &level, NULL);
+    g_print("%g\n", level);
   }
 
-  if (error != NULL) {
-    g_printerr("An error occurred: %s\n", error->message);
-    g_clear_error(&error);
-    g_object_unref(player);
-    return 1;
-  }
+  return TRUE;
+}
 
-  g_object_unref(player);
-  return 0;
+static gboolean status (const gchar *value, PlayerctlPlayer *player, GError **error)
+{
+  gchar *state = NULL;
+
+  g_object_get(player, "status", &state, NULL);
+  g_print("%s\n", state ? state : "Not available");
+  g_free(state);
+
+  return TRUE;
 }
 
-static gchar *list_player_names(GError **err)
+static gboolean get_metadata (const gchar *value, PlayerctlPlayer *player, GError **error)
 {
   GError *tmp_error = NULL;
+  gchar *data;
+
+  if (g_strcmp0(value, "artist") == 0)
+    data = playerctl_player_get_artist(player, &tmp_error);
+  else if (g_strcmp0(value, "title") == 0)
+    data = playerctl_player_get_title(player, &tmp_error);
+  else if (g_strcmp0(value, "album") == 0)
+    data = playerctl_player_get_album(player, &tmp_error);
+  else
+    data = playerctl_player_print_metadata_prop(player, value, &tmp_error);
+
+  if (tmp_error) {
+    g_propagate_error(error, tmp_error);
+    return FALSE;
+  }
 
-  GDBusProxy *proxy = g_dbus_proxy_new_for_bus_sync(
-      G_BUS_TYPE_SESSION,
-      G_DBUS_PROXY_FLAGS_NONE,
-      NULL,
-      "org.freedesktop.DBus",
-      "/org/freedesktop/DBus",
-      "org.freedesktop.DBus",
-      NULL,
-      &tmp_error);
+  g_print("%s", data);
+  g_free(data);
 
-  if (tmp_error != NULL) {
-    g_propagate_error(err, tmp_error);
-    return NULL;
+  return TRUE;
+}
+
+struct PlayerCommand {
+  const gchar *name;
+  gboolean (*func) (const gchar *value, PlayerctlPlayer *player, GError **error);
```

Comment:
Prior to my changes, the main method in this file was 176 lines long, containing a complicated if/else construct of 117 lines. I extracted each piece of functionality into its own function, making each function easier to comprehend (allowing me to detect several memory leaks) and improving the maintainability of the code. All of the changes I made follow good software engineering practices.

I disagree that introducing a table of function pointers makes the control flow *more* complicated; I merely moved the complexity from a 100+ lines if/else to separate functions with a function table. In my opinion, that is easier to follow *and* to extend.

I may have grown the file, but I have not introduced much new code (as you said; my changes "constitute a derivative work which are not substantially different enough to be protected under copyright"). The increase in size is mostly due to having to declare all the now separate functions, and the function table.

I'm not sure if you intended this, but I sense some annoyance in your last sentence. If you do not appreciate my work, simply tell me and I will stop wasting your and my own time. If you felt I was being offensive, then I apologize.

That said, I am happy to make the change you suggested.